File: typescript/planet-time/src/time.ts
/**
 * time.ts — Planet time functions for @interplanet/time
 */

import type { Planet, PlanetTime, MTC, MarsLocalTime } from './types.js';
import { PLANETS, MARS_EPOCH_MS, MARS_SOL_MS } from './constants.js';

const DOW_NAMES  = ['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday'];
const DOW_SHORT  = ['Mon','Tue','Wed','Thu','Fri','Sat','Sun'];

function pad(n: number): string {
  return String(n).padStart(2, '0');
}

/**
 * Get the current time on a planet.
 *
 * @param planet       planet key (e.g. 'earth', 'mars')
 * @param utcMs        UTC milliseconds since Unix epoch
 * @param tzOffsetHours  zone offset in local planet hours (default: 0)
 */
export function getPlanetTime(
  planet: Planet,
  utcMs: number,
  tzOffsetHours = 0,
): PlanetTime {
  // Moon uses Earth's data
  const key = planet === 'moon' ? 'earth' : planet;
  const p   = PLANETS[key as Planet];

  const elapsedMs   = utcMs - p.epochMs + (tzOffsetHours / 24) * p.solarDayMs;
  const totalDays   = elapsedMs / p.solarDayMs;
  const dayNumber   = Math.floor(totalDays);
  const dayFraction = totalDays - dayNumber;

  const localHour = dayFraction * 24;
  const h = Math.floor(localHour);
  const m = Math.floor((localHour - h) * 60);
  const s = Math.floor(((localHour - h) * 60 - m) * 60);

  const { daysPerPeriod, periodsPerWeek, workPeriodsPerWeek } = p;
  let periodInWeek: number;
  let isWorkPeriod: boolean;
  let isWorkHour: boolean;
  if (p.earthClockSched) {
    // Earth-clock scheduling: UTC day-of-week + UTC hour (Mercury and Venus)
    const utcDay = Math.floor(utcMs / 86_400_000);
    const dow = ((utcDay % 7) + 3 + 7) % 7;  // Mon=0..Sun=6
    isWorkPeriod = dow < workPeriodsPerWeek;
    const msOfDay = utcMs - utcDay * 86_400_000;
    const utcHour = Math.floor(msOfDay / 3_600_000);
    isWorkHour = isWorkPeriod && utcHour >= p.workHoursStart && utcHour < p.workHoursEnd;
    periodInWeek = dow;
  } else {
    const totalPeriods = totalDays / daysPerPeriod;
    periodInWeek = ((Math.floor(totalPeriods) % periodsPerWeek) + periodsPerWeek) % periodsPerWeek;
    isWorkPeriod = periodInWeek < workPeriodsPerWeek;
    isWorkHour   = isWorkPeriod && localHour >= p.workHoursStart && localHour < p.workHoursEnd;
  }

  const yearLenDays = p.siderealYrMs / p.solarDayMs;
  const yearNumber  = Math.floor(totalDays / yearLenDays);
  const dayInYear   = totalDays - yearNumber * yearLenDays;

  let solInfo = null;
  if (key === 'mars') {
    const solsPerYear = PLANETS.mars.siderealYrMs / PLANETS.mars.solarDayMs;
    solInfo = { solInYear: Math.floor(dayInYear), solsPerYear: Math.round(solsPerYear) };
  }

  const dowIndex = periodInWeek % 7;

  const data = planet === 'moon' ? PLANETS.moon : p;

  return {
    planet: data.name,
    symbol: data.symbol,
    hour: h, minute: m, second: s, localHour, dayFraction,
    dayNumber, dayInYear: Math.floor(dayInYear), yearNumber, solInfo,
    periodInWeek, isWorkPeriod, isWorkHour,
    dowName: DOW_NAMES[dowIndex]!, dowShort: DOW_SHORT[dowIndex]!,
    solarDayMs: p.solarDayMs, daysPerPeriod, periodsPerWeek, workPeriodsPerWeek,
    timeString: `${pad(h)}:${pad(m)}`,
    timeStringFull: `${pad(h)}:${pad(m)}:${pad(s)}`,
  };
}

/**
 * Get Mars Coordinated Time (MTC).
 *
 * @param utcMs  UTC milliseconds since Unix epoch
 */
export function getMTC(utcMs: number): MTC {
  const totalSols = (utcMs - MARS_EPOCH_MS) / MARS_SOL_MS;
  const sol  = Math.floor(totalSols);
  const frac = totalSols - sol;
  const h = Math.floor(frac * 24);
  const m = Math.floor((frac * 24 - h) * 60);
  const s = Math.floor(((frac * 24 - h) * 60 - m) * 60);
  return { sol, hour: h, minute: m, second: s, mtcString: `${pad(h)}:${pad(m)}` };
}

/**
 * Get Mars local time at a given zone offset.
 *
 * @param utcMs        UTC milliseconds since Unix epoch
 * @param offsetHours  Mars local hours relative to AMT (prime meridian)
 */
export function getMarsTimeAtOffset(utcMs: number, offsetHours: number): MarsLocalTime {
  const mtc = getMTC(utcMs);
  let h = mtc.hour + offsetHours;
  let solDelta = 0;
  if (h >= 24) { h -= 24; solDelta =  1; }
  if (h <   0) { h += 24; solDelta = -1; }
  const fh = Math.floor(h);
  return {
    sol: mtc.sol + solDelta,
    hour: fh, minute: mtc.minute, second: mtc.second,
    timeString: `${pad(fh)}:${pad(mtc.minute)}`,
    offsetHours,
  };
}

File: python/planet-time/src/interplanet_time/_time.py
"""
_time.py — Planet time and MTC calculations.
Exact port of planet-time.js getPlanetTime / getMTC functions.
All functions accept utc_ms (int, milliseconds since Unix epoch).
"""

from __future__ import annotations
import math
from datetime import datetime, timezone

from ._constants import Planet, PLANETS, MARS_EPOCH_MS, MARS_SOL_MS
from ._models import PlanetTime, MTC

# Maps Planet enum → 3-letter timezone zone prefix (mirrors ZONE_PREFIXES in planet-time.js)
_ZONE_PREFIXES: dict[Planet, str] = {
    Planet.MARS:    "AMT",
    Planet.MOON:    "LMT",
    Planet.MERCURY: "MMT",
    Planet.VENUS:   "VMT",
    Planet.JUPITER: "JMT",
    Planet.SATURN:  "SMT",
    Planet.URANUS:  "UMT",
    Planet.NEPTUNE: "NMT",
}

__all__ = ["get_planet_time", "get_mtc", "get_mars_time_at_offset"]


def get_planet_time(
    planet: Planet,
    utc_ms: int,
    tz_offset_h: float = 0,
) -> PlanetTime:
    """Get the current time on a planet.

    Parameters
    ----------
    planet:      Planet enum member
    utc_ms:      UTC milliseconds since Unix epoch
    tz_offset_h: local-hour zone offset from planet prime meridian (default 0)

    Notes
    -----
    Moon maps to Earth's solar day (tidally locked; work schedules on Earth time).
    Mars populates sol_in_year / sols_per_year; all others get None.
    """
    # Moon uses Earth's planetary data
    key = Planet.EARTH if planet == Planet.MOON else planet
    p   = PLANETS[key]

    elapsed_ms  = utc_ms - p['epochMs'] + tz_offset_h / 24 * p['solarDayMs']
    total_days  = elapsed_ms / p['solarDayMs']
    day_number  = math.floor(total_days)
    day_frac    = total_days - day_number

    local_hour  = day_frac * 24
    h           = math.floor(local_hour)
    m           = math.floor((local_hour - h) * 60)
    s           = math.floor(((local_hour - h) * 60 - m) * 60)

    days_per    = p['daysPerPeriod']
    per_week    = p['periodsPerWeek']
    work_per    = p['workPeriodsPerWeek']

    if p.get('earthClockSched'):
        # Mercury and Venus: schedule on UTC Mon–Fri 09:00–17:00
        dt_utc = datetime.fromtimestamp(utc_ms / 1000, tz=timezone.utc)
        period_in_week = dt_utc.weekday()  # 0=Mon..6=Sun
        is_work_period = period_in_week < work_per
        utc_hour = dt_utc.hour + dt_utc.minute / 60.0 + dt_utc.second / 3600.0
        is_work_hour = is_work_period and p['workHoursStart'] <= utc_hour < p['workHoursEnd']
    else:
        total_periods = total_days / days_per
        period_in_week = (
            (math.floor(total_periods) % per_week) + per_week
        ) % per_week
        is_work_period = period_in_week < work_per
        is_work_hour   = (
            is_work_period
            and p['workHoursStart'] <= local_hour < p['workHoursEnd']
        )

    year_len    = p['siderealYrMs'] / p['solarDayMs']
    year_number = math.floor(total_days / year_len)
    day_in_year = total_days - year_number * year_len

    sol_in_year  = None
    sols_per_year = None
    if planet == Planet.MARS:
        sols_per_year_f = PLANETS[Planet.MARS]['siderealYrMs'] / PLANETS[Planet.MARS]['solarDayMs']
        sol_in_year   = math.floor(day_in_year)
        sols_per_year = round(sols_per_year_f)

    prefix = _ZONE_PREFIXES.get(planet)
    if prefix is not None:
        off_sign = "+" if tz_offset_h >= 0 else ""
        zone_id: str | None = f"{prefix}{off_sign}{math.trunc(tz_offset_h)}"
    else:
        zone_id = None

    return PlanetTime(
        hour=h, minute=m, second=s,
        local_hour=local_hour,
        day_fraction=day_frac,
        day_number=day_number,
        day_in_year=math.floor(day_in_year),
        year_number=year_number,
        period_in_week=int(period_in_week),
        is_work_period=bool(is_work_period),
        is_work_hour=bool(is_work_hour),
        time_str=f"{h:02d}:{m:02d}",
        time_str_full=f"{h:02d}:{m:02d}:{s:02d}",
        sol_in_year=sol_in_year,
        sols_per_year=sols_per_year,
        zone_id=zone_id,
    )


def get_mtc(utc_ms: int) -> MTC:
    """Get Mars Coordinated Time (MTC) — the Martian equivalent of UTC."""
    total_sols = (utc_ms - MARS_EPOCH_MS) / MARS_SOL_MS
    sol  = math.floor(total_sols)
    frac = total_sols - sol
    h    = math.floor(frac * 24)
    m    = math.floor((frac * 24 - h) * 60)
    s    = math.floor(((frac * 24 - h) * 60 - m) * 60)
    return MTC(
        sol=sol, hour=h, minute=m, second=s,
        mtc_str=f"{h:02d}:{m:02d}",
    )


def get_mars_time_at_offset(utc_ms: int, offset_h: float) -> PlanetTime:
    """Get Mars local time at a given zone offset (Mars local hours from AMT)."""
    return get_planet_time(Planet.MARS, utc_ms, tz_offset_h=offset_h)
